text/0035-opentelemetry-protocol.md

```diff
@@ -82,23 +82,76 @@ If the client is shutting down (e.g. when the containing process wants to exit)
+When using gRPC transport success response is returned via `ExportResponse` message.
 
-- `FailedRetryable` - processing of telemetry data failed. The client SHOULD record the error and may retry exporting the same data immediately. This can happen when the server is temporarily unable to process the data.
+When an error is returned by the server it falls into 2 broad categories: retryable and not-retryable.
+
+- Retryable errors indicate that processing of telemetry data failed and the client SHOULD record the error and may retry exporting the same data. This can happen when the server is temporarily unable to process the data.
+
+- Not-retryable errors indicate that processing of telemetry data failed and the client MUST NOT retry sending the same telemetry data. The telemetry data MUST be dropped. This for example can happen when the request contains bad data and cannot be deserialized or otherwise processed by the server. The client SHOULD maintain a counter of such dropped data.
+
+When using gRPC transport the server SHOULD indicate retryable errors using code [Unavailable](https://godoc.org/google.golang.org/grpc/codes) and MAY supply additional [details via status](https://godoc.org/google.golang.org/grpc/status#Status.WithDetails) using [RetryInfo](https://github.com/googleapis/googleapis/blob/6a8c7914d1b79bd832b5157a09a9332e8cbd16d4/google/rpc/error_details.proto#L40) containing 0 value of RetryDelay. Here is a sample Go code to illustrate:
+
+```go
+  // Do this on server side.
+  st, err := status.New(codes.Unavailable, "Server is unavailable").
+    WithDetails(&errdetails.RetryInfo{RetryDelay: &duration.Duration{Seconds: 0}})
+  if err != nil {
+    log.Fatal(err)
+  }
+
+  return st.Err()
+```
+
+To indicate not-retryable errors the server is recommended to use code [InvalidArgument](https://godoc.org/google.golang.org/grpc/codes) and MAY supply additional [details via status](https://godoc.org/google.golang.org/grpc/status#Status.WithDetails) using [BadRequest](https://github.com/googleapis/googleapis/blob/6a8c7914d1b79bd832b5157a09a9332e8cbd16d4/google/rpc/error_details.proto#L119). Other gRPC status code may be used if it is more appropriate. Here is a sample Go code to illustrate:
+
+```go
+  // Do this on server side.
+  st, err := status.New(codes.InvalidArgument, "Invalid Argument").
+    WithDetails(&errdetails.BadRequest{})
+  if err != nil {
+    log.Fatal(err)
+  }
+
+  return st.Err()
+```
+
+The server MAY use other gRPC codes to indicate retryable and not-retryable errors if those other gRPC codes are more appropriate for a particular erroneous situation. The client SHOULD interpret gRPC status codes as retryable or not-retryable according to the following table:
+
+|gRPC Code|Retryable?|
+|---------|----------|
+|CANCELLED|Yes|
+|UNKNOWN|No|
+|INVALID_ARGUMENT|No|
+|DEADLINE_EXCEEDED|Yes|
+|NOT_FOUND|No|
+|ALREADY_EXISTS|No|
+|PERMISSION_DENIED|Yes|
+|UNAUTHENTICATED|Yes|
```

Comment:
Should these PERMISSION_DENIED and UNAUTHENTICATED be retryable? It is clear that you need to change the request otherwise this error will happen again, so probably I wouldn't say that the request is retryable.

Reply:
If we make this non-retryable then a misconfiguration (e.g. wrong auth token) will result in OTLP dropping all data that we attempt to send. Hard to tell what is the best approach here. Do we want the sender to keep accumulating the data and retrying it until someone notices the problem and fixes the auth? Or do we want the sender to begin dropping data until someone notices the problem and fixes the auth? I have a weak preferences for retrying in this case, but happy to hear other opinions.

Reply:
I suggest adding a note ("[1] Request ought to be changed to proper credentials", etc) to make clear this is a retryable error that **requires** a change.

Reply:
There is already a link to https://godoc.org/google.golang.org/grpc/codes which explains in details what each of the codes means. PERMISSION_DENIED do not necessarily require the request to be changed, it may also be a server misconfiguration. I think the link to grpc codes provides sufficient information and hits the right spot of being precise enough and not too verbose. I could copy all that text here but I think it is unnecessary.

Reply:
When we define "Retryable" above, we say "processing of *telemetry data* failed". So if we are splitting hairs, we could say PERMISSION_DENIED is retryable because it failed due to auth-z, not "telemetry data". However, if I think of a naive implementation of the sender, such as one that's given an application token to insert into headers (as opposed to using time-varying protocol like OAUTH2 where token might get refreshed after a while), then retrying would result in absolutely the same total request that will again fail.

Similarly, the server that cannot auth-z the client could be thinking "oh, but if you restart me with a different config, then perhaps this request will succeed", which also strikes me as a stretch.

So I would classify both of these as Not retryable. My 2c.

Reply:
Makes sense, I will mark as non-retryable for now. We can refine this in the future if there is any other evidence, since this is a behavioral change and is not a breaking change on the protocol wire format.